healthchain/pipeline/basepipeline.py

```diff
@@ -0,0 +1,463 @@
+
+        if component is None:
+            return wrapper
+        if isinstance(component, BaseComponent):
+            return wrapper(component)
+        elif isinstance(component, Component):
+            return wrapper(component.__call__)
+        else:
+            return wrapper(component)
+
+    def _add_component_at_position(self, new_component, position, reference):
+        """
+        Add a new component to the pipeline at a specified position.
+
+        Args:
+            new_component (PipelineNode): The new component to be added to the pipeline.
+            position (str): The position where the component should be added.
+                            Valid values are 'first', 'last', 'after', 'before', or 'default'.
+            reference (str, optional): The name of the reference component when using 'after' or 'before' positions.
+
+        Raises:
+            ValueError: If an invalid position is provided or if a reference is required but not provided.
+
+        This method handles the insertion of a new component into the pipeline based on the specified position:
+        - 'first': Inserts the component at the beginning of the pipeline.
+        - 'last' or 'default': Appends the component to the end of the pipeline.
+        - 'after' or 'before': Inserts the component relative to a reference component.
+
+        For 'after' and 'before' positions, a reference component name must be provided.
+        """
+        if position == "first":
+            self._components.insert(0, new_component)
+        elif position in ["last", "default"]:
+            self._components.append(new_component)
+        elif position in ["after", "before"]:
+            if not reference:
+                raise ValueError(
+                    f"Reference must be provided for position '{position}'."
+                )
+            offset = 1 if position == "after" else 0
+            self._insert_relative_position(new_component, reference, offset)
+        else:
+            raise ValueError(
+                f"Invalid position '{position}'. Must be 'first', 'last', 'after', 'before', or 'default'."
+            )
+
+    def _insert_relative_position(self, component, reference, offset):
+        """
+        Insert a component relative to a reference component in the pipeline.
+
+        Args:
+            component (PipelineNode): The component to be inserted.
+            reference (str): The name of the reference component.
+            offset (int): The offset from the reference component (0 for before, 1 for after).
+
+        Raises:
+            ValueError: If the reference component is not found in the pipeline.
+        """
+        ref_index = next(
+            (i for i, c in enumerate(self._components) if c.name == reference), None
+        )
+        if ref_index is None:
+            raise ValueError(f"Reference component '{reference}' not found.")
+
+        self._components.insert(ref_index + offset, component)
```

Comment:
I could have missed it but what happens if I add a stage with the name of a stage which exists already? Can i have duplicates? Or should stage names be forced to be unique?

Reply:
it would just be appended to the existing stage, no?